src/main/java/com/microsoft/sqlserver/jdbc/KerbAuthentication.java

```diff
@@ -140,18 +139,34 @@ private void intAuthInit() throws SQLServerException {
             }
             else {
                 Subject currentSubject = null;
+                KerbCallback callback = new KerbCallback(con);
                 try {
                     AccessControlContext context = AccessController.getContext();
                     currentSubject = Subject.getSubject(context);
                     if (null == currentSubject) {
-                        lc = new LoginContext(CONFIGNAME);
+                        lc = new LoginContext(CONFIGNAME, callback);
                         lc.login();
                         // per documentation LoginContext will instantiate a new subject.
                         currentSubject = lc.getSubject();
                     }
                 }
                 catch (LoginException le) {
-                    con.terminate(SQLServerException.DRIVER_ERROR_NONE, SQLServerException.getErrString("R_integratedAuthenticationFailed"), le);
+                    if (authLogger.isLoggable(Level.FINE)) {
+                        authLogger.fine(toString() + "Failed to login using Kerberos due to " + le.getClass().getName() + ":" + le.getMessage());
+                    }
+                    try {
+                        // Not very clean since it raises an Exception, but we are sure we are cleaning well everything
+                        con.terminate(SQLServerException.DRIVER_ERROR_NONE, SQLServerException.getErrString("R_integratedAuthenticationFailed"), le);
+                    } catch (SQLServerException alwaysTriggered) {
+                        String message = String.format("%s due to %s (%s)", alwaysTriggered.getMessage(), le.getClass().getName(), le.getMessage());
+                        if (callback.getUsernameRequested() != null) {
+                            message = String.format("Login failed for Kerberos principal '%s'. %s", callback.getUsernameRequested(), message);
+                        }
```

Comment:
To be consistent with the existing code, can you add this new exception error string in SQLServerResource.java?